book/src/07_join.md

```diff
@@ -0,0 +1,130 @@
+# Joining components
+
+In the last chapter, we learned how to access resources using `SystemData`.
+To access our components with it, we can just fetch a `ReadStorage` and use
+`Storage::get` to get the component associated to an entity. This works quite
+well if you want to access a single component, but what if you want to
+iterate over many components? Maybe some required, others optional, maybe even
+exclude component groups that have a certain component? If we wanted to do
+that using `Storage::get`, the code would become very ugly. So instead
+we worked out a way to conveniently specify that. And this concept is
+known as "Joining".
+
+## Basic joining
+
+We've already seen some basic examples in the last chapters, for
+example how to join over two storages:
+
+```rust,ignore
+for (pos, vel) in (&mut pos_storage, &vel_storage).join() {
+    *pos += *vel;
+}
+```
+
+This simply iterates over the position and velocity components of
+all entities that have both these components. That means all the
+specified components are **required**.
+
+Sometimes, you want to not only get the components of entities,
+but also the entity itself. To do that, we can simply join over
+`&EntitiesRes`. We can get `&EntitiesRes` by de-referencing `Entities`
+and re-referencing the returned value.
+
+```rust,ignore
+for (ent, pos, vel) in (&*entities, &mut pos_storage, &vel_storage).join() {
+    println!("Processing entitiy: {:?}", ent);
+    *pos += *vel;
+}
+```
+
+## Optional components
+
+If we iterate over the `&EntitiesRes` as shown above, we can simply
+use the returned `Entitiy` values to get components from storages as usual.
+
+```rust,ignore
+for (ent, pos, vel) in (&*entities, &mut pos_storage, &vel_storage).join() {
+    println!("Processing entitiy: {:?}", ent);
+    *pos += *vel;
+    
+    let mass: Option<&mut Mass> = mass_storage.get_mut(ent);
+    if let Some(mass) = mass {
+        let x = *vel / 300_000_000.0;
+        let y = 1 - x * x;
+        let y = y.sqrt();
+        mass.current = mass.constant / y;
+    }
+}
+```
+
+In this example we iterate over all entities with a position and a velocity
+and perform the calculation for the new position as usual.
+However, in case the entity has a mass, we also calculate the current
+mass based on the velocity. Thus, mass is an **optional** component here.
+
+## Excluding components
+
+If you want to filter your selection by excluding all entities
+with a certain component type, you can use the not operator (`!`)
+on respective component storage. Its return value is a unit (`()`).
```

Comment:
s/Its return value is a unit/When iterating over the join with negated storage instead of the component we get the unit/